In [2]:
import pandas as pd
import numpy as np

#Import data from the clean file 
df = pd.read_csv('../data/metadata_clean.csv')

#Print the head of the cleaned DataFrame
df.head()

title                              genres  runtime  \
0                    Toy Story   ['Animation', 'Comedy', 'Family']     81.0   
1                      Jumanji  ['Adventure', 'Fantasy', 'Family']    104.0   
2             Grumpier Old Men               ['Romance', 'Comedy']    101.0   
3            Waiting to Exhale      ['Comedy', 'Drama', 'Romance']    127.0   
4  Father of the Bride Part II                          ['Comedy']    106.0   

   vote_average  vote_count  year  
0           7.7      5415.0  1995  
1           6.9      2413.0  1995  
2           6.5        92.0  1995  
3           6.1        34.0  1995  
4           5.7       173.0  1995

## Plot description-based recommender

1. 모델 구축에 필요한 데이터 획득
2. 모든 영화의 플롯 설명(또는 overwise)에 대한 TF-IDF 벡터 만들기
3. 모든 영화의 pairwise similarity score 계산
4. 영화 제목을 argument로 하고 줄거리를 바탕으로 가장 유사한 영화를 출력하는 함수 작성

In [3]:
#Import the original file
orig_df = pd.read_csv('../data/movies_metadata.csv', low_memory=False)

#Add the useful features into the cleaned dataframe
df['overview'], df['id'] = orig_df['overview'], orig_df['id']

df.head()

title                              genres  runtime  \
0                    Toy Story   ['Animation', 'Comedy', 'Family']     81.0   
1                      Jumanji  ['Adventure', 'Fantasy', 'Family']    104.0   
2             Grumpier Old Men               ['Romance', 'Comedy']    101.0   
3            Waiting to Exhale      ['Comedy', 'Drama', 'Romance']    127.0   
4  Father of the Bride Part II                          ['Comedy']    106.0   

   vote_average  vote_count  year  \
0           7.7      5415.0  1995   
1           6.9      2413.0  1995   
2           6.5        92.0  1995   
3           6.1        34.0  1995   
4           5.7       173.0  1995   

                                            overview     id  
0  Led by Woody, Andy's toys live happily in his ...    862  
1  When siblings Judy and Peter discover an encha...   8844  
2  A family wedding reignites the ancient feud be...  15602  
3  Cheated on, mistreated and stepped on, the wom...  31357  
4  Just when George Banks has recovered from his ...  11862

## Creating the TF-IDF matrix

In [4]:
#Import TfIdfVectorizer from the scikit-learn library
from sklearn.feature_extraction.text import TfidfVectorizer

#Define a TF-IDF Vectorizer Object. Remove all english stopwords
tfidf = TfidfVectorizer(stop_words='english')

#Replace NaN with an empty string
df['overview'] = df['overview'].fillna('')

#Construct the required TF-IDF matrix by applying the fit_transform method on the overview feature
tfidf_matrix = tfidf.fit_transform(df['overview'])

#Output the shape of tfidf_matrix
tfidf_matrix.shape

(45466, 75827)

모든 영화의 개요에 대해 75,827차원 벡터를 생성

## Computing the cosine similarity score

모든 영화의 pairwise cosine similarity를 계산. </br>
즉, $i$ 번째 행과 $j$ 번째 열에 있는 셀은 영화 $i$와 $j$ 사이의 유사도 점수를 나타내는 45,466 × 45,466 행렬을 만들 것.

In [5]:
# Import linear_kernel to compute the dot product
from sklearn.metrics.pairwise import linear_kernel

# Compute the cosine similarity matrix
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

## Building the recommender function

In [6]:
#Construct a reverse mapping of indices and movie titles, and drop duplicate titles, if any
indices = pd.Series(df.index, index=df['title']).drop_duplicates()

1. 영화 제목을 argument로 선언
2. index 역방향 mapping으로 영화의 index를 가져옴
3. cosie_sim을 사용하는 모든 영화의 코사인 유사도 점수를 가져온다. (첫 번째 요소가 위치, 두번째 요소가 유사도 점수)
4. 첫 번째를 제외한 상위 10개 - 첫 번째 요소는 자신과의 유사성 점수를 의미하므로 무시

In [7]:
# Function that takes in movie title as input and gives recommendations 
def content_recommender(title, cosine_sim=cosine_sim, df=df, indices=indices):
    # Obtain the index of the movie that matches the title
    idx = indices[title]

    # Get the pairwsie similarity scores of all movies with that movie
    # And convert it into a list of tuples as described above
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Sort the movies based on the cosine similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar movies. Ignore the first movie.
    sim_scores = sim_scores[1:11]

    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]

    # Return the top 10 most similar movies
    return df['title'].iloc[movie_indices]

In [8]:
#Get recommendations for The Lion King - 라이온 킹과 유사한 영화의 추천
content_recommender('The Lion King')

34682    How the Lion Cub and the Turtle Sang a Song
9353                                The Lion King 1½
9115                  The Lion King 2: Simba's Pride
42829                                           Prey
25654                                 Fearless Fagan
17041                                   African Cats
27933              Massaï, les guerriers de la pluie
6094                                       Born Free
37409                                     Sour Grape
3203                                The Waiting Game
Name: title, dtype: object

대부분의 영화가 사자와 연관되어 있거나 라이온 킹을 사랑하는 사람들은 디즈니 영화에 흥미를 가질 가능서잉 높다. <br>
또한, 애니메이션 영화를 보는 것을 선호할 수 있다. </br>

하지만, 이 방법으로는 모든 정보를 포착할 수 없음. </br>

-> metadata를 활용한 추천 (특정 감독, 배우, 하위 장르 등에 대한 개인의 취향을 파악하는 것이 더 효과적)

## Metadata Based Recommender

- genre
- director
- 주요 출연진 3명
- 하위 장르 또는 키워드

In [10]:
# Load the keywords and credits files
cred_df = pd.read_csv('../data/credits.csv')
key_df = pd.read_csv('../data/keywords.csv')

#Print the head of the credit dataframe
cred_df.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

In [11]:
#Print the head of the keywords dataframe
key_df.head()

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

In [12]:
#Convert the IDs of df into int
df['id'] = df['id'].astype('int')

ValueError: invalid literal for int() with base 10: '1997-08-20'

In [15]:
df[df['id']=='1997-08-20']


# 앞의 코드를 실행하면 ValueError가 발생
# -> ID 중 int로 바꿀 수 없는 column이 존재
# -> 제거해야함.

title                                             genres  runtime  \
19730   NaN  ['Carousel Productions', 'Vision View Entertai...      NaN   

       vote_average  vote_count  year  overview          id  
19730           NaN         NaN     0  Released  1997-08-20

In [16]:
# Function to convert all non-integer IDs to NaN
def clean_ids(x):
    try:
        return int(x)
    except:
        return np.nan

#Clean the ids of df
df['id'] = df['id'].apply(clean_ids)

#Filter all rows that have a null ID
df = df[df['id'].notnull()]

In [17]:
# Convert IDs into integer
df['id'] = df['id'].astype('int')
key_df['id'] = key_df['id'].astype('int')
cred_df['id'] = cred_df['id'].astype('int')

# Merge keywords and credits into your main metadata dataframe
df = df.merge(cred_df, on='id')
df = df.merge(key_df, on='id')

#Display the head of the merged df
df.head()

/var/folders/zp/147pq5010bg5vyj070ml25lr0000gn/T/ipykernel_92395/3018193927.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['id'] = df['id'].astype('int')


title                              genres  runtime  \
0                    Toy Story   ['Animation', 'Comedy', 'Family']     81.0   
1                      Jumanji  ['Adventure', 'Fantasy', 'Family']    104.0   
2             Grumpier Old Men               ['Romance', 'Comedy']    101.0   
3            Waiting to Exhale      ['Comedy', 'Drama', 'Romance']    127.0   
4  Father of the Bride Part II                          ['Comedy']    106.0   

   vote_average  vote_count  year  \
0           7.7      5415.0  1995   
1           6.9      2413.0  1995   
2           6.5        92.0  1995   
3           6.1        34.0  1995   
4           5.7       173.0  1995   

                                            overview     id  \
0  Led by Woody, Andy's toys live happily in his ...    862   
1  When siblings Judy and Peter discover an encha...   8844   
2  A family wedding reignites the ancient feud be...  15602   
3  Cheated on, mistreated and stepped on, the wom...  31357   
4  Just when George Banks has recovered from his ...  11862   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...   
3  [{'credit_id': '52fe44779251416c91011acb', 'de...   
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...   

                                            keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...  
2  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...  
3  [{'id': 818, 'name': 'based on novel'}, {'id':...  
4  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

## Wrangling keywords, cast, and crew

- 상위 3개의 키워드만 포함
- 감독만 추출 (나머지 제작진 무시)

In [18]:
# Convert the stringified objects into the native python objects
from ast import literal_eval

features = ['cast', 'crew', 'keywords', 'genres']
for feature in features:
    df[feature] = df[feature].apply(literal_eval)

In [19]:
#Print the first cast member of the first movie in df
df.iloc[0]['crew'][0]

{'credit_id': '52fe4284c3a36847f8024f49',
 'department': 'Directing',
 'gender': 2,
 'id': 7879,
 'job': 'Director',
 'name': 'John Lasseter',
 'profile_path': '/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg'}

In [20]:
# Extract the director's name. If director is not listed, return NaN
def get_director(x):
    for crew_member in x:
        if crew_member['job'] == 'Director':
            return crew_member['name']
    return np.nan

In [21]:
#Define the new director feature
df['director'] = df['crew'].apply(get_director)

#Print the directors of the first five movies
df['director'].head()

0      John Lasseter
1       Joe Johnston
2      Howard Deutch
3    Forest Whitaker
4      Charles Shyer
Name: director, dtype: object

In [27]:
# Returns the list top 3 elements or entire list; whichever is more.
# 영화마다 상위 3개의 장르만 고려
def generate_list(x):
    if isinstance(x, list):
        names = [ele['name'] for ele in x]
        #Check if more than 3 elements exist. If yes, return only first three. 
        #If no, return entire list.
        if len(names) > 3:
            names = names[:3]
        return names

    #Return empty list in case of missing/malformed data
    return []

In [23]:
#Apply the generate_list function to cast and keywords
df['cast'] = df['cast'].apply(generate_list)
df['keywords'] = df['keywords'].apply(generate_list)

#Only consider a maximum of 3 genres
df['genres'] = df['genres'].apply(lambda x: x[:3])

In [24]:
# Print the new features of the first 5 movies along with title
df[['title', 'cast', 'director', 'keywords', 'genres']].head(3)

title                                            cast  \
0         Toy Story             [Tom Hanks, Tim Allen, Don Rickles]   
1           Jumanji  [Robin Williams, Jonathan Hyde, Kirsten Dunst]   
2  Grumpier Old Men      [Walter Matthau, Jack Lemmon, Ann-Margret]   

        director                                           keywords  \
0  John Lasseter                               [jealousy, toy, boy]   
1   Joe Johnston  [board game, disappearance, based on children'...   
2  Howard Deutch       [fishing, best friend, duringcreditsstinger]   

                         genres  
0   [Animation, Comedy, Family]  
1  [Adventure, Fantasy, Family]  
2             [Romance, Comedy]

문서 벡터를 만들기 위해 vectorizer 사용 <br>
만약 두 배우의 이름이 같다면 (ex. Ryan Reynolds 과 Ryan Gosling), vectorizer는 분명히 다른 개체이지만 두 명을 동일하기 취급 <br>

-> 따라서 키워드와 배우 및 감독 이름 사이의 **공백을 제거하고 모두 소문자로 변환** <br>
ex. ryangosling과 ryanreynolds

In [25]:
# Function to sanitize data to prevent ambiguity. 
# Removes spaces and converts to lowercase
def sanitize(x):
    if isinstance(x, list):
        #Strip spaces and convert to lowercase
        return [str.lower(i.replace(" ", "")) for i in x]
    else:
        #Check if director exists. If not, return empty string
        if isinstance(x, str):
            return str.lower(x.replace(" ", ""))
        else:
            return ''

In [26]:
#Apply the generate_list function to cast, keywords, director and genres
for feature in ['cast', 'director', 'genres', 'keywords']:
    df[feature] = df[feature].apply(sanitize)

## Creating the metadata soup

In [28]:
#Function that creates a soup out of the desired metadata
def create_soup(x):
    return ' '.join(x['keywords']) + ' ' + ' '.join(x['cast']) + ' ' + x['director'] + ' ' + ' '.join(x['genres'])

In [29]:
# Create the new soup feature
df['soup'] = df.apply(create_soup, axis=1)

In [30]:
#Display the soup of the first movie
df.iloc[0]['soup']

'jealousy toy boy tomhanks timallen donrickles johnlasseter animation comedy family'

## Generating the recommendations

TF-IDF Vectorizer를 사용하는 대신 Count Vectorizer를 사용할 예정
- TF-IDF Vectorize를 사용하면 상대적으로 많은 수의 영화에서 연기와 감돌을 한 배우와 더 적은 weight를 부여할 수 있기 때문
- 더 많은 영화에 출연하거나 감독했다고해서 패널티를 주고 싶지 않았기 때문

In [34]:
# Import CountVectorizer
from sklearn.feature_extraction.text import CountVectorizer

#Define a new CountVectorizer object and create vectors for the soup
count = CountVectorizer(stop_words='english')
count_matrix = count.fit_transform(df['soup'])

In [35]:
#Import cosine_similarity function
from sklearn.metrics.pairwise import cosine_similarity

#Compute the cosine similarity score (equivalent to dot product for tf-idf vectors)
cosine_sim2 = cosine_similarity(count_matrix, count_matrix)

In [36]:
# Reset index of your df and construct reverse mapping again
df = df.reset_index()
indices2 = pd.Series(df.index, index=df['title'])

In [37]:
content_recommender('The Lion King', cosine_sim2, df, indices2)

29607                                          Cheburashka
40904                   VeggieTales: Josh and the Big Wall
40913    VeggieTales: Minnesota Cuke and the Search for...
27768                                 The Little Matchgirl
15209             Spiderman: The Ultimate Villain Showdown
16613                            Cirque du Soleil: Varekai
24654                                  The Seventh Brother
29198                                      Superstar Goofy
30244                                              My Love
31179                Pokémon: Arceus and the Jewel of Life
Name: title, dtype: object

- 사자 뿐만 아니라 다른 정보를 포착할 수 있었는데, 대부분의 영화는 애니메이션이고 의인화된 캐릭터를 특징으로 한다. <br>
- 'Pokémon: Arceus and the Jewel of Life'과 'The Lion King' 모두 몇 년 후 죄를 지은 사람에게 복수하기 위해 돌아오는 만화 의인화된 인물 등장